### PR TITLE
Create win_powershell utility

### PR DESCRIPTION
**Pull request by twangboy**
### What does this PR do?

Creates a win_powershell utility with one function (so far) that will detect if a module is installed on the system without opening PowerShell. This cleans up the debug log where the lazy loader used to run PowerShell to detect modules... which was costly and verbose.

Modified win_servermanager and win_pki to use this utility.

Modified cmd.shell_info to use this utility.

Used https://github.com/saltstack/salt/pull/36890
### What issues does this PR fix or reference?

Related to: https://github.com/saltstack/salt/issues/36252
### Previous Behavior

Modules that worked with PowerShell would try to detect if a PowerShell module was installed using PowerShell with was costly and verbose in the logs.
### New Behavior

Now it looks in the PSModulePath for the module instead of trying to import with PowerShell and failing... much faster.
### Tests written?

No

### CODE REVIEW

**Comment by twangboy** [comment]
Go Go Jenkins!

**Comment by damon-atkins** [comment]
Note this other PR #36890 Add available PS module names to shell_info
One would assume #36890 is no longer required. But the discussion and code might be worth looking at.

I suggest its just called powershell instead of win_powershell as powershell is available on Unix/Linux, and it should support both.

**Comment by damon-atkins** [comment]
I would also suggest you use cmd.shell_info for this check
```
def __virtual__():
    if not salt.utils.is_windows():
        return (False, 'Failed to load win_powershell: '
                       'The utility only works on Windows systems.')
    return __virtualname__
```

**Comment by damon-atkins** [comment]
On second thoughts `def __virtual__():`  it should just return true.  And just document its the callers responsibility to call cmd.shell_info to check if powershell is installed.

For example module_exists will still work, as the check is outside of powershell.

For example win_servermanager.py should call is_windows, then  cmd.shell_info then module_exists.
Assuming 'ServerManager' is not available on Unix/Linux.

**Comment by m03** [comment]
Modules may be nested into subdirectories, which this should be modified to account for. For an example, try the following in PowerShell:

```
PS C:\> $Env:PsModulePath
C:\Users\Administrator\Documents\WindowsPowerShell\Modules;C:\Program Files\WindowsPowerShell\Modules;C:\Windows\system32\WindowsPowerShell\v1.0\Modules\

PS C:\> New-Item -Path 'C:\Program Files\WindowsPowerShell\Modules\TestParent1\TestParent2\TestParent3\TestModule\testmodule.psd1' -ItemType File -Force

PS C:\> Get-Module -ListAvailable | Where { $_.Name -eq 'TestModule' }

    Directory: C:\Program Files\WindowsPowerShell\Modules\TestParent1\TestParent2\TestParent3

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Manifest   0.0        TestModule
```

Should also make sure to support .cdxml, .dll, .psm1, and .xaml, in addition to .psd1.

**Comment by damon-atkins** [comment]
On UNIX/Linux `PSModulePath` may not be set. The default path is:
'{1}/.local/share/powershell/Modules;/opt/microsoft/powershell/{2}/Modules'.format(os.environ.get('HOME',''),ret['version_raw'])

**Comment by cachedout** [comment]
@twangboy Are you able to address the final concern from @damon-atkins ?

**Comment by damon-atkins** [comment]
PowerShell  CentOS, Ubuntu and Mac is available from https://github.com/PowerShell/PowerShell

**Comment by damon-atkins** [comment]
Re: cmd.shell_info()
`ret['availablemodules'] = salt.utils.powershell.get_modules()`
Please change to have an _ (underscore) between words.
My preference if this is included to be called
`modules_available` as it matches version_raw.  
`ret['modules_available'] = salt.utils.powershell.get_modules()`
or
`ret['modules'] = salt.utils.powershell.get_modules()`

From my point of view this is optional.  As I would expect that anyone doing anything with powershell will import this new utils module.

If we include `ret['modules'] = salt.utils.powershell.get_modules()`
i.e. shell_info returns a list of modules.
Then in the long term we should do the same for python, perl  and maybe zsh and others.

**Given shell_info is called inside def virtual in each module, it may have to much of a performance hit to included it** 

Alternative is
```
def shell_info(shell,**kwargs):
    - cut code -- note the location as this will pick up powershell on mac and unix/linux

    else:
        ret['installed'] = None  # Have an unexpect result

    if shell == 'powershell' and ret['installed'] and salt.utils.is_true(kwargs.get('module_list', False)):
       ret['modules'] = salt.utils.powershell.get_modules()
```

**Comment by damon-atkins** [comment]
Re: win_pki.py and win_servermanager.py
```
   powershell_info = __salt__['cmd.shell_info']('powershell')
   if not powershell_info['installed'] or not salt.utils.powershell.module_exists('ServerManager'):
        return False, 'Failed to load win_servermanager module: ' \
                      'ServerManager module not available. ' \
                      'May need to install Remote Server Administration Tools.'
```
or
```
   powershell_info = __salt__['cmd.shell_info']('powershell',modules_list=True)
   if not powershell_info['installed'] or not 'ServerManager' in powershell_info['modules']):
        return False, 'Failed to load win_servermanager module: ' \
                      'ServerManager module not available. ' \
                      'May need to install Remote Server Administration Tools.'
```

**Comment by twangboy** [comment]
@cachedout This should address the issues now...

**Comment by damon-atkins** [comment]
ps_version vs psversion    The idea was to match the variable names coming from $PSVersionTable, but in all lowercase (because python is case sensitive).    So that if someone did run $PSVersionTable  they would get the same result/variables. The _ (underscore) was more about the ones we have added as extra.

```
PS C:\ > $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.14393.206
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14393.206
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

The code takes any keys within the reg and adds it so if microsoft  come along and add more keys into the reg they will turn up as is.

At the end it does not worry me. ps_version vs ps_version etc, as long as powershell people will understand.

**Comment by damon-atkins** [comment]
@twangboy   thanks for adding the all the code for determine modules it should have a significant elapse time on windows.

**Comment by damon-atkins** [comment]
@cachedout  Out of interest, whats the story with 80 char  line wrapping.   I thought it was try for 80 but 100 is ok if it makes the code easy to read.

**Comment by cachedout** [comment]
@damon-atkins It's as you've described the style guide. We prefer vertical code but not at the expense of readability.

**Comment by cachedout** [comment]
@damon-atkins Are you happy with where this is now or are there any outstanding concerns?

**Comment by damon-atkins** [comment]
See comment about "PS C:\ > $PSVersionTable"

**Comment by damon-atkins** [comment]
1.  See `PS C:\ > $PSVersionTable` comment.  For SaltStack to decide, I am not really a powershell programmer.

2. Re: powershell.utils.
shell_info() will return the powershell version on Unix/Linux/Mac if powershell is in path

I would write this:
```
    # See if PSModulePath is defined in the system environment
    if env_var not in os.environ:
        # Once we determine how to find the PSVersion outside of powershell in
        # other OS'es (non Windows), we'll add additional paths here
        default_paths = [
            '{0}/.local/share/powershell/Modules'.format(os.environ.get('HOME')),
            '/usr/local/share/powershell/Modules']

        # Check if defaults exist, add them if they do
        root_paths = []
        for item in default_paths:
            if os.path.exists(item):
                root_paths.append(item)

        # Did we find any, if not log the error and return
        if not root_paths:
            log.error('Environment variable not present: %s', env_var)
            log.error('Default paths not found')
            return ret

    else:
        # Use PSModulePaths
        root_paths = [
            str(path) for path in os.environ[env_var].split(';') if path]

```
As this (double check the indentation)
```
if is_windows():
  ps_module_path=os.environ.get('PSModulePath',
     '{0}\system32\WindowsPowerShell\v1.0\Modules\;{1}\WindowsPowerShell\Modules\'.format(os.environ.get('SystemRoot'),os.environ.get('ProgramFiles')))
else:
  # Should not be calling this if powershell is not installed so no checks on __salt__['cmd.shell_info']('powershell')['version_raw']
  # the following path is based on CentOS, assume all Unix/Linux/Mac are the same.
  ps_module_path=os.environ.get('PSModulePath',
     '{0}/.local/share/powershell/Modules;/opt/microsoft/powershell/{1}/Modules'.format(os.environ.get('HOME',''),__salt__['cmd.shell_info']('powershell')['version_raw']))


   root_paths = []
   for item in ps_module_path.split(';'):
      if os.path.exists(item):
          root_paths.append(item)
```

Everything else is fine.

**Comment by damon-atkins** [comment]
Most likely need \\ for windows paths or r'fkdlskfl'

**Comment by twangboy** [comment]
@damon-atkins @cachedout Getting the powershell version for the additional paths is going to have to happen in another PR
We can't use __salt__ from within the salt.util. We'll have to create win_reg salt util and create another version of the cmd.shell_info function within the powershell module to get the powershell version.
Also, I don't know how to get the powershell version in linux outside of using powershell.
I started doing this but it ended up being more complicated than I anticipated. I recommend getting this in for now and building on it in another PR. It may need to go into develop.

**Comment by damon-atkins** [comment]
> get the PowerShell version in Linux outside of using PowerShell.

Thats why the code calls powershell directly on Linux. PowerShell on Linux is under a second.  Powershell on windows is about 5 seconds. Therefore not worth writing code to speed it up Linux as its already fast.   But the module code should be able to work on all platforms.

__salt__['cmd.shell_info']('powershell')['version_raw'] no loop
__salt__['cmd.shell_info']('powershell',list_modules=True)['version_raw'] never ending loop

I Agree we need to get this in. So it can get in the next Carbon RC.
You can leave the Unix support for modules out for now.

**Comment by damon-atkins** [comment]
If I can find some spare time I will raise another PR... I do not want hold up the next RC, because Unix/Linux support is not on the critical path.

File: salt/modules/cmdmod.py [hunk 1]
```diff
@@ -2653,16 +2654,25 @@ def shells():
     return ret
 
 
-def shell_info(shell):
+def shell_info(shell, **kwargs):
```
Comment (cachedout) [inline salt/modules/cmdmod.py]: We do not typically accept `**kwargs` in execution modules. Why is this necessary here?

File: salt/modules/cmdmod.py [hunk 2]
```diff
@@ -2653,16 +2654,25 @@ def shells():
...
 
-    :param str shell: Name of the shell. Support shells/script languages include
+    Args:
+        shell (str): Name of the shell. Support shells/script languages include
         bash, cmd, perl, php, powershell, python, ruby and zsh
-    :return: Properties of the shell specifically its and other information if available.
-    :rtype: dict
+
+    Kwargs:
+        list_modules (bool): True to list modules available to the shell.
```
Comment (cachedout) [inline salt/modules/cmdmod.py]: Why is this in `**kwargs` and not just a normal keyword argument?

File: salt/utils/powershell.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        '{0}\\WindowsPowerShell\\v1.0\\Modules\\'.format(system_dir),
+        '{0}\\WindowsPowerShell\\Modules'.format(program_files)]
+    default_paths = ';'.join(default_paths)
+
+    ps_module_path = os.environ.get('PSModulePath', default_paths)
+
+    # Check if defaults exist, add them if they do
+    ps_module_path = ps_module_path.split(';')
+    for item in ps_module_path:
+        print(item)
```
Comment (cachedout) [inline salt/utils/powershell.py]: Please remove this.

File: salt/modules/cmdmod.py (hunk 1 above)
Comment (cachedout) [inline salt/modules/cmdmod.py]: We've been actively engaged in trying to correct cases where `**kwargs` is used in other execution modules.

Comment (damon-atkins) [inline salt/modules/cmdmod.py]: If that is the case you have 633 module def's to go.  I think I would be the other way around, as your fixing the API's parameters location.  Salt needs to be more fluid than this and `**kwargs` allows it to be fluid.

Comment (damon-atkins) [inline salt/modules/cmdmod.py]: See #36991 put your thoughts here please.
